test/test_buffer_protocol.py

```diff
@@ -0,0 +1,175 @@
+
+# Tests for the `frombuffer` function (only work on CPU):
+#   Constructs tensors from Python objects that implement the buffer protocol,
+#   without copying data.
+class TestBufferProtocol(common.TestCase):
+    def _run_test(self, shape, dtype, count=-1, first=0, offset=None, **kwargs):
+        numpy_dtype = common.torch_to_numpy_dtype_dict[dtype]
+
+        if offset is None:
+            offset = first * get_dtype_size(dtype)
+
+        numpy_original = common.make_tensor(shape, torch.device("cpu"), dtype).numpy()
+        original = memoryview(numpy_original)
+        # First call PyTorch's version in case of errors.
+        # If this call exits successfully, the NumPy version must also do so.
+        torch_frombuffer = torch.frombuffer(original, dtype=dtype, count=count, offset=offset, **kwargs)
+        numpy_frombuffer = numpy.frombuffer(original, dtype=numpy_dtype, count=count, offset=offset)
+
+        self.assertEqual(numpy_frombuffer, torch_frombuffer)
+        self.assertEqual(numpy_frombuffer.__array_interface__["data"][0], torch_frombuffer.data_ptr())
+        return (numpy_original, torch_frombuffer)
+
+    @onlyCPU
+    @dtypes(*common.torch_to_numpy_dtype_dict.keys())
+    def test_same_type(self, device, dtype):
+        self._run_test((), dtype)
+        self._run_test((4,), dtype)
+        self._run_test((10, 10), dtype)
+
+    @onlyCPU
+    @dtypes(*common.torch_to_numpy_dtype_dict.keys())
+    def test_requires_grad(self, device, dtype):
+        def _run_test_and_check_grad(requires_grad, *args, **kwargs):
+            kwargs["requires_grad"] = requires_grad
+            _, tensor = self._run_test(*args, **kwargs)
+            self.assertTrue(tensor.requires_grad == requires_grad)
+
+        if dtype.is_floating_point or dtype.is_complex:
+            _run_test_and_check_grad(True, (), dtype)
+            _run_test_and_check_grad(True, (4,), dtype)
+            _run_test_and_check_grad(True, (10, 10), dtype)
+        _run_test_and_check_grad(False, (), dtype)
+        _run_test_and_check_grad(False, (4,), dtype)
+        _run_test_and_check_grad(False, (10, 10), dtype)
+
+    @onlyCPU
+    @dtypes(*common.torch_to_numpy_dtype_dict.keys())
+    def test_with_offset(self, device, dtype):
+        # Offset should be valid whenever there is, at least,
+        # one remaining element
+        for i in range(SIZE):
+            self._run_test(SHAPE, dtype, first=i)
+
+    @onlyCPU
+    @dtypes(*common.torch_to_numpy_dtype_dict.keys())
+    def test_with_count(self, device, dtype):
+        # Count should be valid for any valid in the interval
+        # [-1, len(input)], except for 0
+        for i in range(-1, SIZE + 1):
+            if i != 0:
+                self._run_test(SHAPE, dtype, count=i)
+
+    @onlyCPU
+    @dtypes(*common.torch_to_numpy_dtype_dict.keys())
+    def test_with_count_and_offset(self, device, dtype):
+        # Explicit default count [-1, 1, 2, ..., len]
+        for i in range(-1, SIZE + 1):
```

Comment:
This is fine and doesn't need to change, but in the future it might be easier to just handwrite some sample inputs for offset, count, and their combination, instead of exhaustively enumerating all possible combinations